pilot/pkg/networking/core/v1alpha3/cluster.go

```diff
@@ -452,6 +452,13 @@ func applyUpstreamTLSSettings(cluster *v2.Cluster, tls *networking.TLSSettings)
 			},
 			Sni: tls.Sni,
 		}
+
+		// Upgrade to H2 only if we know other side is envoy via ISTIO_MUTUAL
```

Comment:
Nit: `s/envoy/part of the service mesh/` - it doesn't have to be Envoy, any Istio-blessed proxy should do ;)